internal/api/attestationconfigapi/cli/client/reporter.go

```diff
@@ -0,0 +1,358 @@
+		apiObject: versions,
+		signer:    c.signer,
+	})
+
+	return executeAllCmds(ctx, c.s3Client, ops)
+}
+
+// addVersionToCache adds the given version to the cache.
+func (c Client) addVersionToCache(ctx context.Context, variant variant.Variant, inputVersion any, date time.Time) error {
+	dateStr := date.Format(VersionFormat) + ".json"
+	obj := apiVersionObject{version: dateStr, variant: variant, cached: true}
+	obj.setVersion(inputVersion)
+	cmd := putCmd{
+		apiObject: obj,
+		signer:    c.signer,
+	}
+	return cmd.Execute(ctx, c.s3Client)
+}
+
+// findMinVersion returns the minimal version in the cache among the last cacheWindowSize versions.
+func (c Client) findMinVersion(
+	ctx context.Context, attestationVariant variant.Variant, versionDates []string,
+) (any, string, error) {
+	var getMinimalVersion func() (any, string, error)
+
+	actionForVariant(attestationVariant,
+		func() {
+			getMinimalVersion = func() (any, string, error) {
+				return findMinimalVersion[attestationconfigapi.TDXVersion](ctx, attestationVariant, versionDates, c.s3Client, c.cacheWindowSize)
+			}
+		},
+		func() {
+			getMinimalVersion = func() (any, string, error) {
+				return findMinimalVersion[attestationconfigapi.SEVSNPVersion](ctx, attestationVariant, versionDates, c.s3Client, c.cacheWindowSize)
+			}
+		},
+	)
+	return getMinimalVersion()
+}
+
+func findMinimalVersion[T attestationconfigapi.TDXVersion | attestationconfigapi.SEVSNPVersion](
+	ctx context.Context, variant variant.Variant, versionDates []string,
+	s3Client *client.Client, cacheWindowSize int,
+) (T, string, error) {
+	var minimalVersion *T
+	var minimalDate string
+	sort.Sort(sort.Reverse(sort.StringSlice(versionDates))) // sort in reverse order to slice the latest versions
+	versionDates = versionDates[:cacheWindowSize]
+	sort.Strings(versionDates) // sort with oldest first to to take the minimal version with the oldest date
+
+	for _, date := range versionDates {
+		obj, err := client.Fetch(ctx, s3Client, apiVersionObject{version: date + ".json", variant: variant, cached: true})
+		if err != nil {
+			return *new(T), "", fmt.Errorf("get object: %w", err)
+		}
+		obj.variant = variant // variant is not set by Fetch, set it manually
+
+		if minimalVersion == nil {
+			v := obj.getVersion().(T)
+			minimalVersion = &v
+			minimalDate = date
+			continue
+		}
+
+		// If the current minimal version has newer versions than the one we just fetched,
+		// update the minimal version to the older version.
+		if isInputNewerThanOtherVersion(variant, *minimalVersion, obj.getVersion()) {
+			v := obj.getVersion().(T)
+			minimalVersion = &v
+			minimalDate = date
+		}
+	}
+
+	return *minimalVersion, minimalDate, nil
+}
+
+func isInputNewerThanOtherVersion(variant variant.Variant, inputVersion, otherVersion any) bool {
+	var result bool
+	actionForVariant(variant,
+		func() {
+			input := inputVersion.(attestationconfigapi.TDXVersion)
+			other := otherVersion.(attestationconfigapi.TDXVersion)
```

Comment:
Are we sure at this point that the versions can't have the wrong type?

Reply:
Yes, if this panics the code is incorrect.